### DISCUSSION TITLE
Robot-Inventor: Activate recovery mode Linux - SecurityError: Access denied.

### DISCUSSION THREAD

**Discussion by scotus-49** [pybricks/support#294] [Q&A] [answered]
I am trying to use the recovery mode to rescue my Robot Inventor Hub. I have followed the extra instructions for Linux, and downloaded and copied the udev rules files to /etc/udev/rules.d/99-pybricksdev.rules. 

I am still getting the SecurityError: Access denied message. Can you help?

**Comment by laurensvalk** [comment]
Which browser do you use? I had the same with Chromium on a more recent Linux build.

It did work in Google Chrome, I think.

But since you're on Linux, you can also do:

```
sudo apt install dfu-util
dfu-util -a 0 -s 0x08008000 -D /path/to/backup.bin
```

**Comment by scotus-49** [comment]
Thanks for the help. I was using Chromium, and after installing Google Chrome it seemed to work. I got the following messages:

Preparing installation.
Installing firmware.
Starting hub.
Done!

Then the Bluetooth button went dark, and the Hub is completely unresponsive - playing dead. I used the file 96112deb24a934bfc19a13c7ea620f54.bin {storm - v1.0.06.0034-b0c335b - 2.1.4.13} downloaded from the linked site. What should I do now?

**Comment by laurensvalk** [comment]
What is the size of file you downloaded? ~100Kb by any chance? I think I might have just done the same as you, right-clicking and save-as. But that gives you an HTML file. Despite being called `.bin`, it's not the .bin file.

So be sure to click on the file and then click download, which gives you [this link](https://github.com/gpdaniels/spike-prime/raw/master/firmware/storm%20-%20v1.0.06.0034-b0c335b%20-%202.1.4.13/96112deb24a934bfc19a13c7ea620f54.bin).

That file is about 500kb. Just run the procedure again with the new file.

Let me know how it goes.

**Comment by scotus-49** [comment]
Thanks again Laurens - it worked this time! I should have seen that the icon of my downloaded file was showing that it was a text file - about 115K.

Now to my windows machine to reload the Pybricks using the Lego app...

**Comment by laurensvalk** [comment, marked as answer]
Or, try this page so you don't have to use Windows :wink: 

https://install.pybricks.com/

**Comment by scotus-49** [comment]
Thanks for the link - I've used it to install the firmware, and it all went fine.  

However, with my newly installed firmware, I'm having problems connecting to Pybricks Code. Whether I use Windows or Linux - I'm getting:
"Connected to hub but failed to get Device Information dervice.
Ensure that you are using the most recent firmware..."

I am assuming that in Pybricks mode the InventorHub bluetooth button is inactive.

**Comment by laurensvalk** [comment]
Yeah, we're doing quite extensive Bluetooth revisions at the moment. Hopefully most of those will be fixed this week.

I'll keep this issue open as a reminder that I upload the latest firmware to this installation page.

**Comment by scotus-49** [comment]
Thanks for the update. I thought I'd done something wrong again...

Should I again use your page https://install.pybricks.com/ to update the firmware, or should I wait a bit?

**Reply by laurensvalk** [reply]
I've just updated that test page with the latest version of the firmware.

One day we're hoping to integrate this in the regular editor.

**Reply by scotus-49** [reply]
I've just flashed it onto my InventorHub, and the process worked fine, with the hub rebooting into the Lego firmware. However, even with the updates to beta12 of Pybricks Code, the error messages I mentioned above are still there. Neither windows (10 Pro) or my Linux (Ubuntu 20.04) manages to connect to my InventorHub running Pybricks. I tried  Google Chrome, Chromium and Edge, and all had the same large red warning box. 

However, my InventorHub running the standard firmware connects over Bluetooth with the Lego app on Win 10 without a problem.

**Reply by laurensvalk** [reply]
I wonder if your browser might still have the old firmware version cached. You could try hitting CTRL+F5 on the firmware installation page before installing again.

I think I could have a look at this on Friday.

This should all be taken care of when we integrate it in the main app in a future release :)

**Reply by scotus-49** [reply]
SUCCESS (with Windows 10)

I followed your suggestion, and ran <CTRL>+F5 in the browser (Linux, Google Chrome) before reinstalling. The installation worked fine, but that red warning box still came up when trying to connect by Bluetooth to the hub while running Pybricks. 

I decided to try my Win 10 box using Google Chrome to do a re-reinstallation. All went well, and I loaded a simple motor script into the Chrome app running Pybricks Code. Pairing was no problem, and I soon had a whirring motor sound cheering me up.

Here's the console messages from Win 10:
```
script.ts:311 Currently running dual-boot firmware.
script.ts:529 {
    "version": "v1.0.06.0034-b0c335",
    "checksum": 1231192444,
    "size": 557056,
    "vector": 134765257
}

/firmware-backup-v1.0.06.0034-b0c335.bin:1 Failed to load resource: the server responded with a status of 404 ()
script.ts:432 Checksum read from flash: 1231192444 and calculated from binary: 1231192444
script.ts:477 {size: 971508, checksum: 3988942305, vector: 135074709, version: "v1.0.06.0034-b0c335+pybricks"}
script.ts:494 Bytes to copy: 971508
script.ts:516 {appl_checksum: 1231192444, new_appl_image_stored_checksum: 3988942305, appl_calc_checksum: 1812799996, new_appl_valid: true, new_appl_image_calc_checksum: 3988942305, …}
beta.pybricks.com/:1 Uncaught (in promise) DOMException: The device has been lost.
```
Under "script.ts.494" my linux version gave 963792 Bytes to be copied, versus 971508 in Windows.

I'll check things a bit more thoroughly later, but meanwhile, a big heart-felt thankyou to you both for all your hard work and effort.

Greetings from Berlin,
Francis Sim
